### ISSUE TITLE
YAML anchor tags aren't deep merged anymore

### ISSUE THREAD

**Issue by xtreme-sameer-vohra** [concourse/concourse#424]
On v1.2.0 we noticed that YAML anchor tags are shallow merged. As of v1.0.0 they were being deep merged.

To illustrate, here is an example 

```
top:
  config: &config
  	key1: value1
  	map1: 
  	  mapkey1: value1
  	  mapkey2: value2

  otherconfig:
  	<<: *config
  	map1:
  	  mapkey1: some-other-value
```
here's what we get on v1.0.0

```
top:
  config: &config
  	key1: value1
  	map1: 
  	  mapkey1: value1
  	  mapkey2: value2

  otherconfig:
  	key1: value1
  	map1:
  	  mapkey1: some-other-value
  	  mapkey2: value2
```

Here's what we're getting on v1.2.0
```
top:
  config: &config
  	key1: value1
  	map1: 
  	  mapkey1: value1
  	  mapkey2: value2

  otherconfig:
  	key1: value1
  	map1:
  	  mapkey1: some-other-value
```   

We use anchor tags extensively to cut down on the duplication in our YAML files.

**Comment by jfrank-pivotal** [comment]
We hit this as well.

For now, we're using YAML like this with nested anchors (because `params` isn't being deep merged correctly):

```yaml
templates:
  - &setup-config
    task: setup-config
    params: &setup-config-params
      ACCESS_TOKEN: {{digitalocean-access-token}}

      SSH_USER: {{digitalocean-ssh-user}}
      SSH_KEY: {{digitalocean-ssh-key}}

      REGION: sfo1
      IMAGE: ""
      SIZE: 1gb
      PRIVATE_NETWORKING: 'false'
      BACKUPS_ENABLED: 'false'
      IP6: 'false'
      OS: ""
```

which we reference like

```yaml
jobs:
- name: digitalocean-centos-7

- name: digitalocean-ubuntu-xenial
  serial: true
  plan:
  - get: bin-rc
    params:
      globs:
      - concourse_linux_amd64

  - <<: *setup-config
    params:
      <<: *setup-config-params
      IMAGE: ubuntu-16-04-x64
      OS: ubuntu
```

**Comment by Pirolf** [comment]
We spent a while looking into this.  In short, the golang YAML library is actually behaving as expected and we don't plan to fix this.

More specifically, we used to pass a specific struct (`atc.Config`) in for `yaml.Unmarshal` to hyrdate, but we had to change it to an interface for compatibility with the new version semantics. The golang implementation of `Unmarshal` does deep merges only when a struct is provided, otherwise fields from the anchor are not propagated when the anchor merge takes place.  It's somewhat unclear which interpretation is correct, but there's an open issue on the golang YAML library referencing this behavior here: https://github.com/go-yaml/yaml/issues/81.

While we agree that supporting deep merges would help reduce repeated configuration code, it seems that YAML containing such deep merges may not actually be supported, and we only enjoyed that functionality in the past due to a difference of interpretation between the YAML spec and the golang implementation.  In Ruby, we saw the same issue as in Go:
```
{
	"config" => {
		"key1" => "value1", 
		"map1" => {
			"mapkey1" => "value1", "mapkey2" => "value2"
		}
	}, "otherconfig" => {
		"key1" => "value1", 
		"map1" => {
			"mapkey1" => "some-other-value"
		}
	}
}
```
